### DISCUSSION TITLE
Общее обсуждение  2023

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2_classic#114] [General]
..

**Comment by Kor-Angar** [comment]
в dia_garond_backinkap4_info  не проверки на то выполнен ли квест Ангара?

```D
    if (DJG_ANGARGOTAMULETT == false)
    {
        AI_Teleport(djg_angar,"OW_DJG_WATCH_STONEHENGE_01");
        b_startotherroutine(djg_angar,"Start");
    };
    DJG_ANGAR_SENTTOSTONES = TRUE;
```

**Comment by Kor-Angar** [comment]
у Ли можно было скрыть вариант о покупке доспехов для ОнД

**Reply by Kor-Angar** [reply]
```D
func int dia_lee_armorh_condition()
{
	//if((KAPITEL == 3) && ((other.guild == GIL_SLD) || (other.guild == GIL_DJG)))
	if ((KAPITEL == 3) && (other.guild == GIL_SLD))
	{
		return TRUE;
	};
};


func int dia_lee_buyarmorh_condition()
{
    if (hero.guild == GIL_DJG)  { return FALSE;	};
    if (LEE_SLDHGIVEN == TRUE)	{ return FALSE;	};
    
	if (Npc_KnowsInfo(other,dia_lee_armorh))
	{
		return TRUE;
	};
};
```

**Comment by Kor-Angar** [comment]
@dosinabox  такой вот баг сегодня у меня был в оригинальных скриптах акеллы https://www.youtube.com/watch?v=Osp9Cn0bD44

**Reply by dosinabox** [reply]
а у меня уже исправлено
https://github.com/dosinabox/g2_classic/commit/0a281d89e5bac3002904002108f7c4884d7e098d#diff-0c52f0cca0c12710ee704615dbec437d9dfa8b3a83307f041367bf64a2d2bd63

**Reply by Kor-Angar** [reply]
> Надо будет не забыть проверить это в испансокй версии

а там всё нормально :-)

**Comment by Kor-Angar** [comment]
в b_buildlearnstring other лучше заменить на hero

**Comment by Kor-Angar** [comment]
предлагаю перепроверит точку у Билгота и в классик и в НВ и в неоф. обновление.

![20231229021729_1](https://github.com/dosinabox/g2_classic/assets/49514764/43bcd5d5-a8a4-477d-9609-be86267946fc)
![20231229021748_1](https://github.com/dosinabox/g2_classic/assets/49514764/b437d673-2bac-439a-8ca8-0ab3d69258a7)

**Reply by Kor-Angar** [reply]
а что у него с гильдией? :-)

`const int GIL_OUT						= 12	; 	// Outlander (z.B. kleine Bauernhцfe)`

**Reply by Kor-Angar** [reply]
MDK

```D
func int DIA_Bilgot_BEIBRUECKEANGEKOMMEN_Condition ()
{
	if (Npc_GetDistToWP(self,"START")<8000) 				
	{
		return TRUE;
	};
};
```

**Comment by Kor-Angar** [comment]
надо бы проверить защиту от стрел у всех монстров. в частности у огненного голема.

**Reply by Kor-Angar** [reply]
- [ ] огненный голем
- [ ] скелет маг

**Reply by Kor-Angar** [reply]
а у меня в связи с этим даже плагин есть :-) https://steamcommunity.com/sharedfiles/filedetails/?id=2984317778

**Comment by Kor-Angar** [comment]
надо еще раз проверить командира орков в 4 главе при игре за пала / мага. мне показалось что у него не было лута.

**Reply by Kor-Angar** [reply]
у скелета мага тоже нету лута

**Reply by Kor-Angar** [reply]
у меня только такая правка по командирам орков стоит


```D
instance ORCELITE_ANTIPALADIN(C_NPC)
{
    ORCELITE_ANTIPALADIN_old();
	b_setvisuals_orcelite();
};

instance ORCELITE_ANTIPALADIN1(C_NPC)
{
    ORCELITE_ANTIPALADIN1_old();
	b_setvisuals_orcelite();
};

instance ORCELITE_ANTIPALADIN2(C_NPC)
{
    ORCELITE_ANTIPALADIN2_old();
    b_setvisuals_orcelite();
};

instance ORCELITE_ANTIPALADIN3(C_NPC)
{
    ORCELITE_ANTIPALADIN3_old();
	b_setvisuals_orcelite();
};
```

**Reply by dosinabox** [reply]
У скелета мага и не должно быть.
Командиры уже исправлены, но исправление ещё не загружалось.

**Reply by Kor-Angar** [reply]
ясно. странно что скелета мага решили лута. я бы в неоф. обновление закинул им после смерти по свитку призыва скелета.

**Comment by Kor-Angar** [comment]
паладину Сенгранту  нехватает чего то для счастья в 4ой главе. или новой инстанции или очистки инвентаря.

**Reply by Kor-Angar** [reply]
```D
func void B_GiveTradeInv(var C_NPC slf)
{

    ...

    if C_NPC2023(slf, PAL_267_Sengrath)
    && (Npc_IsInState	(slf, ZS_Dead))
    {   
        Npc_ClearInventory(slf);
        if (Npc_HasItems(sengrath,itrw_mil_crossbow))			 {   Npc_RemoveInvItem(sengrath,itrw_mil_crossbow);      };
        if (!Npc_HasItems(sengrath,itrw_sengrathsarmbrust_mis))  {   CreateInvItem(sengrath,itrw_sengrathsarmbrust_mis); };
        b_createambientinv(slf);
        B_GiveTradeInv_Sengrath(slf);
        return;
    };  
    ...
}; 
```

**Reply by Kor-Angar** [reply]
![20231230183738_1](https://github.com/dosinabox/g2_classic/assets/49514764/e58106a4-85ca-4bbe-921e-bedf9dda2cd4)

**Comment by Kor-Angar** [comment]
через вторый ворота героя не пускают. герой уже подмастерье с доступом в верхний квартал.

**Reply by Kor-Angar** [reply]
![20240102034750_1](https://github.com/dosinabox/g2_classic/assets/49514764/531a4a8f-2fdb-4385-8549-8545ed32e045)

**Reply by Kor-Angar** [reply]
![20240102011324_1](https://github.com/dosinabox/g2_classic/assets/49514764/e24f231b-8e1e-4b51-9ab7-d83e1b7ac2b2)

**Comment by Kor-Angar** [comment]
func void b_startotherroutine(var C_NPC slf,var string newroutine)

**Reply by Kor-Angar** [reply]
тут отдаётся ссылка на НПС. а  в коде отдаётся то ссылка на инстанцию то сама инстанция.
![image](https://github.com/dosinabox/g2_classic/assets/49514764/6daa0d32-da73-4f69-bc74-f9e55c7d8b52)

**Reply by Kor-Angar** [reply]
можно добавить для таких ситуаций дополнительную функцию

```D
func void B_StartOtherRoutineIntNpc(var int intnpc,var string newroutine)
{
    var C_NPC npc;  npc = Hlp_GetNpc(intnpc);
    
    if(Hlp_IsValidNpc(npc)
    {
        b_startotherroutine(npc,newroutine)
    };
};
```

тогда

```D
			B_StartOtherRoutine (Sekob,"FleeDMT"); 
			B_StartOtherRoutine (Rosi,"FleeDMT"); 
			B_StartOtherRoutine (Till,"FleeDMT");
			if(BalthasarMovedToBengar == FALSE)
			{
				b_startotherroutine(balthasar,"FleeDMT");
			};
			B_StartOtherRoutineIntNpc(BAU_933_Rega,"FleeDMT"); 
			B_StartOtherRoutineIntNpc(BAU_934_Babera,"FleeDMT"); 
			B_StartOtherRoutineIntNpc(BAU_937_BAUER,"FleeDMT"); 
			B_StartOtherRoutineIntNpc(BAU_938_BAUER,"FleeDMT"); 
```

**Reply by Kor-Angar** [reply]
```D
            B_StartOtherRoutine (Malak,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_962_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_964_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_965_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_966_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_967_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_968_Bauer,"FleeFromPass"); 
            B_StartOtherRoutineIntNpc (BAU_969_Bauer,"FleeFromPass"); 
```

**Reply by dosinabox** [reply]
В целом, не критично и работает и так. Лучше добавить несколько нужных людей в b_initnpcglobals, чем создавать новую функцию.

**Reply by Kor-Angar** [reply]
> Лучше добавить несколько нужных людей в b_initnpcglobals, чем создавать новую функцию.

ну это вопрос вкуса. лично мне кажется `b_initnpcglobals` в целом было неудачное решение. от этого проблем сильно больше чем пользы. я бы его выпилил как явление под корень :-D и тогда воцарился бы мир и порядок. путаницы бы не было, а значит и потенциальных ошибок.

а что касается не критично и так работает. это да. не критично и работает. поэтому я даже не стал писать в топик про баги это. а написал это просто что бы зафиксировать такое явление :-)

**Reply by Kor-Angar** [reply]
и ты не просто добавишь пару НПС, а создашь еще для каждого по переменной ;-)

**Comment by Kor-Angar** [comment]
нету метки Bromor_isAlive_Kap3

**Reply by Kor-Angar** [reply]
и fernando_isAlive_Kap3

**Comment by Kor-Angar** [comment]
B_ENTER_NEWWORLD_Kapitel_3 

           B_KillNpc 		(Ulf);

а тут наоборот. надо отдавать инстанцию нпс :-)
то есть 

            B_KillNpc(nov_602_ulf);

**Reply by dosinabox** [reply]
```
b_killnpc(engrom);
b_killnpc(sengrath);
b_killnpc(bruder);
b_killnpc(vlk_leiche3);
b_killnpc(vlk_leiche2);
b_killnpc(strf_leiche1);
b_killnpc(strf_leiche2);
b_killnpc(strf_leiche3);
b_killnpc(strf_leiche4);
b_killnpc(strf_leiche5);
b_killnpc(strf_leiche6);
b_killnpc(strf_leiche7);
b_killnpc(strf_leiche8);
b_killnpc(pal_leiche1);
b_killnpc(pal_leiche2);
b_killnpc(pal_leiche3);
b_killnpc(vlk_leiche1);
b_killnpc(pal_leiche4);
b_killnpc(pal_leiche5);
b_killnpc(olav);
```

**Reply by Kor-Angar** [reply]
я так щепетильно стал относится к ссылки на инстанцию и самой инстанции потому что без этого когда начинаешь что то писать новое вообще не понятно когда что должно быть. но если постоянно обращать внимание где ссылка , а где инстанция постепенно всё становится на свои места. отчасти в этом есть вина пираней, которые вот так делают и сами смешивают одно с другим :-D

**Comment by Kor-Angar** [comment]
Мелдор и Мо не успокаиваются

**Reply by Kor-Angar** [reply]
так же оплата штрафа, не стёрла память стражнику в тяжёлых доспехах

**Comment by Kor-Angar** [comment]
документ "список трав константино" можно сдавать вместе с травами для него

**Comment by Kor-Angar** [comment]
B_CheckDeadMissionNPCs из скриптпатча под классик
[B_CheckDeadMissionNPCs.zip](https://github.com/dosinabox/g2_classic/files/13815228/B_CheckDeadMissionNPCs.zip)

**Comment by Kor-Angar** [comment]
@dosinabox  в zs_dead 


	b_givetradeinv(self);
	b_givedeathinv(self);
	//b_clearruneinv(self); // повторный вызов, уже вызывалась из b_givetradeinv
